### ISSUE TITLE
[New Concept Exercise]: Class Inheritance

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#3096]
This issue describes how to implement the `class-inheritance` concept exercise  **concept exercise** for the Python track.
The related **concept documents improvement issue** can be found [here](https://github.com/exercism/python/issues/3095#issuecomment-1153019285).

<br>

## ✅ Getting started

_If you have not yet created or contributed to a concept exercise, this issue will require some upfront reading to give you the needed background knowledge. Some good example exercises to look at in the repo:_

<details>
<summary>💡<b>Example Exercises</b>💡</summary>
<br>

   1. [Little Sister's Vocabulary](https://github.com/exercism/python/tree/main/exercises/concept/little-sisters-vocab)
   2. [Meltdown Mitigation](https://github.com/exercism/python/tree/main/exercises/concept/meltdown-mitigation) 
   3. [Making the Grade](https://github.com/exercism/python/tree/main/exercises/concept/making-the-grade)
   4. [Ellen's Alien Game](https://github.com/exercism/python/tree/main/exercises/concept/ellens-alien-game)

</details>

We also recommend completing one or more of the concept exercises (_they're called "learning exercises"_) [on the website](https://exercism.org/tracks/python/concepts). 

**`Please please read the docs before starting.`** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please go through the following documents:

**General Contributing Docs:**

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | - [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

**Documents on Language Tracks and Concept Exercises**

- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concept Exercises?](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Exercise Specifications](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Exercise Stories](https://exercism.org/docs/building/tracks/stories)

## 🎯 Goal

This concept exercise is meant to teach an understanding/use of `class-inheritance`  in Python. 

<br>

## 💡Learning objectives

- more fully understand the uses (_and possible abuses_) of `classes`  in Python
- understanding and use of `single inheritance` in Python
- understanding and use of `multiple inheritance` in Python
  - the diamond inheritance problem
  - Pythons MRO (_method resolution order_)
  - pros and cos of `mixins` () to isolate `multiple inheritance` issues
- when and when not to use `super()` to call a parent `class` constructor or method

<br>

## 🤔 Concepts

- `classes`
- `inheritance` in `OOP`
- `single inheritance`
- `multiple inheritance`
- `minxins`
- `super()`

<br>

## 🚫 Topics that are Out of scope

<br>
<details>
<summary><b>Concepts & Subjects that are Out of Scope</b> (<em>click to expand</em>)</summary>
<br>

- `class-composition` as an explicit topic (_it is ok to use limited composition as needed in example code_)
- `dataclasses`
- `decorators` outside of `@property`, `@staticmethod`, and `@classmethod` (_the student should already be exposed to these in the prerequisite exercise_)
- `generators`
- `coroutines`
- `descriptors` (_these will get their own exercise_)
- using a `class` as a decorator
- performance considerations

</details>
<br>

## ↩️  Prerequisites

These are the concepts/concept exercises the student should be familiar with before taking on/learning this concept.  
<br>
<details>
<summary>Prereqs  (<em>click to expand</em>)</summary>

- `basics`
- `booleans`
- `classes`
- `class-customization`
- `comparisons`
- `decorators`
- `dicts`
- `functions`
- `higher-order-functions`
- `iteration`
- `lists`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `tuples`

</details>
<br>

##  📚 Resources for Writing and Reference

<details>
<summary><b>Resources</b> (<em>click to expand</em>)</summary>
<br>

- [classes (Python tutorial)](https://docs.python.org/3/tutorial/classes.html)
- [Real Python: Object-Oriented Programming in Python 3](https://realpython.com/python3-object-oriented-programming/)
- [Real Python: Inheritance and Composition: A Python OOP Guide](https://realpython.com/inheritance-composition-python/)
- [Towards DataScience: Understand Inheritance in Python](https://towardsdatascience.com/understand-inheritance-in-python-74f8e3025f3c) -- (_this is not a good resource for hints or links, since it requires an account signup_)
  - [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle)
- [Multiple Inheritance in Python](https://www.python-course.eu/python3_multiple_inheritance.php)
- [Multiple Inheritance is Hard (Ned Batchelder)](https://nedbatchelder.com/blog/201210/multiple_inheritance_is_hard.html)
- [Python.org: MRO in Python 2.3+](https://www.python.org/download/releases/2.3/mro/)
- [Dig into Python super() and MRO](https://gaopinghuang0.github.io/2018/12/29/dig-into-python-super-and-mro)
- [Multiple inheritance and mixin classes in Python](https://www.thedigitalcatonline.com/blog/2020/03/27/mixin-classes-in-python/)
- [Making Python classes more modular using mixins](https://www.residentmar.io/2019/07/07/python-mixins.html)
- [Write composable, reusable Python classes using Mixinsl)](https://dev.to/bikramjeetsingh/write-composable-reusable-python-classes-using-mixins-6lj)
- [Mixins in Python and Ruby Compared](https://andrewbrookins.com/technology/mixins-in-python-and-ruby-compared/)
- [Python Data Model - Python Docs](https://docs.python.org/3/reference/datamodel.html#the-standard-type-hierarchy)

</details>
<br>  

### Exercise Ideas & Stories

Should you need inspiration for an exercise story, you can find a collection [here](https://exercism.org/docs/building/tracks/stories).  You can also port an exercise from another track, **but please make sure to only to include tasks that actually make sense in Python** and that add value for a student.  Remove/replace/add tasks as needed to make the concept clear/workable.

<br>

##  📁 Exercise Files to Be Created

<details open>
<summary><b>File Detail for this Exercise</b></summary>
<table>
<td>
<br>

* ### Exercise `introduction.md`

  For more information, see [**Exercise** `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docsintroductionmd)

  - This can summarize/paraphrase the linked concept documents if they have already been created (either the `about` or the `introduction`).   The summary does need to have enough information and examples  for the student to complete all the tasks outlined for this concept exercise.

* ### Exercise `instructions.md`

   For more information, see [`instructions.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docsinstructionsmd)

   Instructions for an exercise usually center on a story that sets up the code challenge to be solved.  You can create your own story, or fork one from the ones listed [here](https://exercism.org/docs/building/tracks/stories).  Please make sure to give credit to the original authors if you use a story or fork an exercise.

* ### Exercise `Exemplar.py` Solution
  For more information, see [exemplar implementation](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-exemplar-implementation).
  
  This file should not use syntax or datas structures not introduced in this exercise or in this exercise's prerequisites.  It will be used as an "ideal" solution for the challenge, so make sure it conforms to PEP8 and other formatting conventions, and **does not use single letter variable names**.  It should also include proper module and function-level docstrings.  However, it should **NOT** include typehinting or type aliases.

* ### `<Exercise>.py` (Stub) for Implementation
  For more information, see [stub implementation](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-stub-implementation).
  
  This file should provide the expected function names imported for testing, and optionally TODO comments and or docstrings to aid the student in their implementation.  TODOs and docstrings are not required.

* ### `<Exercise>_Test.py` Files
  For more information, see [Tests](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-tests).
  Additionally, please note that Python associates exercise tasks to tests via a [Pytest Marker](https://docs.pytest.org/en/7.1.x/example/markers.html), and uses [`unittest subtests`](https://docs.python.org/3/library/unittest.html#subtests) as a form of test paramaterization.  See the test file for [`Little Sisters Vocab`](https://github.com/exercism/python/blob/main/exercises/concept/little-sisters-vocab/strings_test.py) for examples of how these techniques work.

* ### Exercise `Hints.md`

  For more information on writing hints see [`hints.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docshintsmd)

  - Hints should provide enough information to get most students "un-stuck" and moving toward a solution.  They should **not** provide a student with a direct solution.
  - You can refer to one or more of the resources linked in this issue above, or analogous resources from a trusted source.  We prefer using links within the  [Python Docs](https://docs.python.org/3/) as the primary go-to, but other resources listed above are also good.  Please try to avoid paid or subscription-based links if possible.

* ### Exercise Metadata Files Under  `.meta/config.json`

  For more information on exercise  `.meta/`  files and formatting, see [concept exercise metadata files](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#metadata-files)

   * `.meta/config.json` - see [this link](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-metaconfigjson) for the fields and formatting of this file.
   * `.meta/design.md` - see [this link](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-metadesignmd) for the formatting of this file.  Please use the **Goal**, **Learning Objectives**,**Concepts**, **Prerequisites** and , **Out of Scope**  sections  from this issue.

<br>
</td>
</table>
</details>
<br>

## ♾️ Exercise Metadata - Track

For more information on concept exercises and formatting for the Python track `config.json` , please see [`config.json`](https://exercism.org/docs/building/tracks/config-json).  The track `config.json` file can be found in the root of the Python repo.

You can use the below for the exercise **UUID**.  You can also generate a new  one via [exercism configlet](https://github.com/exercism/configlet), [uuidgenerator.net](https://www.uuidgenerator.net/version4), or any other favorite method.  The UUID must be a valid  [**V4 UUID**](https://en.wikipedia.org/wiki/Universally_unique_identifier).

- **Exercise UUID** :  ``
- **concepts** should be filled in from the Concepts section in this issue
- **prerequisites** should be filled in from the Prerequisites section in this issue    

<br>

## 🎶 Implementation Notes

- As a reminder, code in the `.meta/examplar.py` file should **only use syntax & concepts introduced in this exercise or one of its prerequisite exercises.**  We run all our `examplar.py` files through PyLint, but do not strictly require module docstrings.  We **do** require function docstrings similar to [PEP257](https://www.python.org/dev/peps/pep-0257/).  See [this concept exercise `exemplar.py`](https://github.com/exercism/python/blob/main/exercises/concept/meltdown-mitigation/.meta/exemplar.py) for an example.
- Please **do not use** comprehensions, generator expressions, or other syntax not previously covered either in the introduction to this exercise, or to one of its prerequisites.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- In General, tests should be written using `unittest.TestCase` and the test file should be named `<EXERCISE-NAME>_test.py`.  

    - All asserts should contain a "user friendly" failure message (_these will display on the webiste to students, so be as clear as you can_). 
    - We use a `PyTest custom mark` to link test cases to exercise task numbers.  
    - We also use `unittest.subtest` to parameterize test input where/when needed.
Here is an [example testfile](https://github.com/exercism/python/blob/main/exercises/concept/making-the-grade/loops_test.py) that shows all three of these in action.

- While we do use [PyTest](https://docs.pytest.org/en/stable/) as our test runner and for some implementation tests, please check with a maintainer before using  a PyTest-specific test method, fixture,  or feature.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

<br>

##  🆘  Next Steps & Getting Help

1.  **`If you'd like to work on this issue, comment saying "I'd like to work on this"`** (_there is no real need to wait for a response, just go ahead, we'll assign you and put a `[claimed]` label on the issue_).
2. If you have any questions while implementing, please post the questions as comments in here, or contact one of the maintainers on our Slack channel.

**Comment by Guillermo-N** [comment]
Hello;
I am working on this, so if you can label it as `[claimed]` and assign it to me will be 👌.

I was thinking about adding some class diagrams to `introduction.md`, I would like to ask if there is any style guideline to follow regarding graphics. Just to fit the looks of it to the rest of the page.

**Comment by BethanyG** [comment]
Hi @Guillermo-N 👋🏽 

Apologies for the delay - I've assigned this issue to you.  Thanks so much for taking it on! ✨ 


> _I was thinking about adding some class diagrams to introduction.md, I would like to ask if there is any style guideline to follow regarding graphics. Just to fit the looks of it to the rest of the page._

We really don't have a style guide for graphics.  In the past (_due to exercises being downloadable to a students machine_), we largely shied away from graphics.  So the overarching advice is to make sure that you don't _solely_ rely on graphics to explain things, since you can't be confident that a student who has downloaded the exercise can see them.  If the student has an IDE that will render a Markdown preview, they will show up in the instruction text.  Otherwise, they will show as links to the GitHub repo where they are stored (_more on that below_).  Here are some other things off the top of my head that I'd recommend:

1.  Make sure to have good [alt text](https://accessibility.huit.harvard.edu/describe-content-images) for each diagram/image.
2. Try to use/save diagrams as `svgs`, since they'll resize much nicer in mobile and smaller-screen context.
3. Choose colors from the surrounding page, but faded or lighter in saturation, so that they don't cause a user's eyes to "bounce" back and forth, or get tired.
4.  While we don't currently have a dark mode for the website, we will have one eventually.  Try to plan graphics so that they will display nicely on both light and dark screens.
5.  Aim for "medium" size.  Don't have a graphic so big it crowds out everything else.  Conversely, don't have a graphic so detailed that it won't be readable on a smaller screen - in other words - don't make a graphic that will require a resize of +400% to read.

For the graphics you do include, you should upload/PR them to the [V3 files repo](https://github.com/exercism/v3-files).  Once they are approved and merged there, you can link to them in your documentation here.  If a graphic is approaching 1MB in size or bigger, you'll want to use the GitHub LFS system to upload them to the repo.  More info on that here:  https://git-lfs.github.com/ 

For the graphics I've included (see this [exercise instructions append](https://github.com/exercism/python/blob/main/exercises/practice/simple-linked-list/.docs/instructions.append.md) and the [exercise hints file](https://github.com/exercism/python/blob/main/exercises/practice/simple-linked-list/.docs/hints.md), I have used [mermaid JS](https://mermaid.live/), and then saved the diagram out as SVG which worked fairly well.  VS Code has extensions that can help with that process.

You can also reach out on our [forum](https://forum.exercism.org/latest) and one of the friendly community members there would be glad to help --  or you can also ping me here on GitHub, should you have more questions or issues.

**Comment by Guillermo-N** [comment]
That is true, I didn't thought about the Markdown not be render. 
Because the graphics are simple I will try to do them in ASCII.

That should be okay, right?

**Comment by Guillermo-N** [comment]
Just submitted what I think can be the a first revision.

**Comment by Guillermo-N** [comment]
Hello @BethanyG;

Any updates or tips about what to change on the code I did summited?

**Comment by BethanyG** [comment]
@Guillermo-N  

Hi.  🙂   Apologies for the delay in getting back to you.  Unfortunately, your submission came in as we were closing this repo to contributions, and putting things on pause.  You can read more about that decision [here](https://exercism.org/blog/freeing-our-maintainers).  While I'd be happy to reevaluate things in a couple of months, I am not going to be able to review your PR right now, and will be closing it with a message.  I am very sorry to have to do this to you, but we do not (right now) have the bandwidth to work on this concept exercise, and need to focus on other exercises and fixes that are higher priority.
